### PR TITLE
[FLINK-19896][table-blink] Improve first-n-row fetching in the Rank o…

### PR DESCRIPTION
**Pull request by tragicjun**
## What is the purpose of the change

Fetching first N rows is a special case of TopN, which currently is supported by the Rank operator. However, the full-fledged TopN implementation is heavyweight in that it involves sort key comparison and raw row buffering. This leads to large state and memory buffer that is not needed for the first-n-row case. This PR introduces a lightweight variant of TopN function to handle it more efficiently. 

## Brief change log

  -  Add `AppendOnlyFirstNFunction` as a variant of AppendOnlyTopNFunction
  -  Modify `StreamExecRank` to instantiate `AppendOnlyFirstNFunciton` when the first-n-row case is determined  

## Verifying this change

This change added tests and can be verified as follows:

  -  Added unit tests in`AppendOnlyFirstNFunctionTest` 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by tragicjun** [comment]
@wuchong @godfreyhe would you mind take a look?

**Comment by JingsongLi** [comment]
Hi @tragicjun, can you rebase latest master?

**Comment by tragicjun** [comment]
> Hi @tragicjun, can you rebase latest master?

Hi @JingsongLi rebase latest master done, please take a look.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyFirstNFunction.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import static org.apache.flink.table.runtime.util.StateConfigUtil.createTtlConfig;
+
+/**
+ * A variant of {@link AppendOnlyTopNFunction} to handle first-n case.
+ *
+ * <p>The input stream should only contain INSERT messages.
+ */
+public class AppendOnlyFirstNFunction extends AbstractTopNFunction {
+
+    private static final long serialVersionUID = -889227691088906246L;
```
Comment (JingsongLi) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyFirstNFunction.java:43]: serialVersionUID  should be 1L at first, From the Flink code style.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecRank.java [hunk 2]
```diff
@@ -207,18 +209,35 @@ public StreamExecRank(
...
-                            sortKeyComparator,
-                            sortKeySelector,
-                            rankType,
-                            rankRange,
-                            generateUpdateBefore,
-                            outputRankNumber,
-                            cacheSize);
+            if (sortFields.length == 1
+                    && inputType.getChildren().get(sortFields[0]).getTypeRoot()
+                            == LogicalTypeRoot.TIMESTAMP_WITHOUT_TIME_ZONE
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecRank.java:214]: Just `TIMESTAMP_WITHOUT_TIME_ZONE`? I don't see any `proc-time`.

**Review by JingsongLi — CHANGES_REQUESTED** [review]
Thanks for the rebasing. @tragicjun Can you create a ITCase or point out which ITCase you want to solve?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyFirstNFunction.java (hunk 1 above)
Comment (tragicjun) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/rank/AppendOnlyFirstNFunction.java:43]: serialVersionUID is at the 2nd in all other classes, thus in consistent with them.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecRank.java (hunk 2 above)
Comment (tragicjun) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecRank.java]: sorry, forgot to push the latest commit.

**Comment by tragicjun** [comment]
> Thanks for the rebasing. @tragicjun Can you create a ITCase or point out which ITCase you want to solve?

Thanks for reviewing. I have added an ITCase as RankJsonPlanITCase.testFirstN(). Please let me know if any issues.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecRank.java [hunk 3]
```diff
@@ -207,18 +211,42 @@ public StreamExecRank(
...
-                            sortKeySelector,
-                            rankType,
-                            rankRange,
-                            generateUpdateBefore,
-                            outputRankNumber,
-                            cacheSize);
+            boolean isAppendOnlyFirstN = false;
+            if (sortFields.length == 1) {
+                LogicalType sortKeyType = inputType.getChildren().get(sortFields[0]);
+                if (sortKeyType instanceof LocalZonedTimestampType
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecRank.java:217]: `TypeCheckUtils.isProcTime`

Comment (tragicjun) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecRank.java]: great, thx

**Comment by JingsongLi** [comment]
@tragicjun still some compile failure.

**Comment by tragicjun** [comment]
> @tragicjun still some compile failure.

sorry, some format violation. Project build passed on my local, please review.

**Review by JingsongLi — APPROVED** [review]
Thanks @tragicjun for the update.
Looks good to me, merging...

**Pull request merged**